libs/gltfio/include/gltfio/NodeManager.h

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GLTFIO_NODEMANAGER_H
+#define GLTFIO_NODEMANAGER_H
+
+#include <filament/FilamentAPI.h>
+
+#include <utils/bitset.h>
+#include <utils/compiler.h>
+#include <utils/CString.h>
+#include <utils/EntityInstance.h>
+#include <utils/FixedCapacityVector.h>
+
+namespace utils {
+class Entity;
+} // namespace utils
+
+namespace gltfio {
+
+class FNodeManager;
+
+/**
+ * NodeManager is used to add annotate entities with glTF-specific information.
+ *
+ * Node components are created by gltfio and exposed to users to allow inspection.
+ *
+ * Nodes do not store the glTF hierarchy or names; see TransformManager and NameComponentManager.
+ */
+class UTILS_PUBLIC NodeManager {
+public:
+    using Instance = utils::EntityInstance<NodeManager>;
+    using Entity = utils::Entity;
+    using CString = utils::CString;
+
+    /**
+     * Returns whether a particular Entity is associated with a component of this NodeManager
+     * @param e An Entity.
+     * @return true if this Entity has a component associated with this manager.
+     */
+    bool hasComponent(Entity e) const noexcept;
+
+    /**
+     * Gets an Instance representing the node component associated with the given Entity.
+     * @param e An Entity.
+     * @return An Instance object, which represents the node component associated with the Entity e.
+     * @note Use Instance::isValid() to make sure the component exists.
+     * @see hasComponent()
+     */
+    Instance getInstance(Entity e) const noexcept;
+
+    /**
+     * Creates a node component and associates it with the given entity.
+     * @param entity            An Entity to associate a node component with.
+     *
+     * If this component already exists on the given entity, it is first destroyed as if
+     * destroy(Entity e) was called.
+     *
+     * @see destroy()
+     */
+    void create(Entity entity);
+
+    /**
+     * Destroys this component from the given entity.
+     * @param e An entity.
+     *
+     * @see create()
+     */
+    void destroy(Entity e) noexcept;
+
+    void setMorphTargetNames(Instance ci, utils::FixedCapacityVector<CString>&& names) noexcept;
+    const utils::FixedCapacityVector<CString>& getMorphTargetNames(Instance ci) const noexcept;
+
+    void setExtras(Instance ci, CString&& extras) noexcept;
```

Comment:
Pass-by-value would create a deep copy of the CString, would it not?  The only setter caller is the loader, which definitely does not need to retain a copy, so I wanted to force a movement to occur.  (This API is public just in case people wanted to use the ECS getters instead of the FilamentAsset bridge.)

`extras` is a glTF term, it's their version of userdata, it's just a string metadata.  We ignored it for a long time until a 3P contribution added support.

Reply:
it would do deep copy only if the caller doesn't use `std::move()` -- but what if the caller wanted to keep its copy of extras?

I'm okay leaving it this way, but it's slightly strange for a public API.

Reply:
Ah, good point, since I'm in control of the loader, I can be smart there while giving users a choice.  I agree with you now.   In fact I think the vector should be pass-by-value too.

Reply:
Fixed.